### PR TITLE
Make SoftwareComponentContainer extensible

### PR DESCRIPTION
**Pull request by jvandort**
Updates DefaultSoftwareComponentContainer to be an extensible polymorphic container.
This will allow our internal plugins and external plugins to register component implementations
so that the component container may instantiate component instances on behalf of the user.

In addition, this PR updates the DefaultJvmSoftwareComponent to be multi-instance aware so that
once the type is made public, users will be able to use the now-extensible software component
container to create arbitrary instances of the default JVM software component

- [x] Add kotlin DSL tests

### CODE REVIEW

**Comment by jvandort** [comment]
@bot-gradle test this

File: subprojects/plugins/src/main/java/org/gradle/jvm/component/internal/DefaultJvmSoftwareComponent.java [hunk 1]
```diff
@@ -218,7 +219,14 @@ private Configuration createApiElements(SourceSet sourceSet, PublishArtifact jar
     }
 
     private Configuration createSourceElements(ConfigurationContainer configurations, ProviderFactory providerFactory, ObjectFactory objectFactory, SourceSet sourceSet) {
-        Configuration variant = configurations.create(SOURCE_ELEMENTS_VARIANT_NAME);
+
+        // TODO: Why are we using this non-standard name? For the `java` component, this
+        // equates to `mainSourceElements` instead of `sourceElements` as one would expect.
+        // Can we change this name without breaking compatibility? Is the variant name part
+        // of the component's API?
+        String variantName = sourceSet.getName() + SOURCE_ELEMENTS_VARIANT_NAME_SUFFIX;
```
Comment (tresat) [inline subprojects/plugins/src/main/java/org/gradle/jvm/component/internal/DefaultJvmSoftwareComponent.java:227]: 👍 This is a good question.  Have we ever removed a consumable variant created by a core plugin before?  What is the deprecation procedure for variants?  We're just changing a configuration name here, but it seems like we would have go through a deprecation cycle.

File: subprojects/plugins/src/main/java/org/gradle/jvm/component/internal/DefaultJvmSoftwareComponent.java [hunk 2]
```diff
@@ -65,21 +63,16 @@
...
-import static org.gradle.api.plugins.JavaPlugin.SOURCES_ELEMENTS_CONFIGURATION_NAME;
 
 /**
  * The software component created by the Java plugin. This component owns the consumable configurations which contribute to
- * this component's variants. Additionally, this component also owns the {@link SourceSet#MAIN_SOURCE_SET_NAME main} {@link SourceSet}
- * and transitively any domain objects which are created by the {@link BasePlugin} on the main source set's behalf. This includes the
- * source set's resolvable configurations and buckets, as well as any associated tasks.
- * <p>
- * This component was written assuming it will only be instantiated once, and therefore it hard-codes the names of the
- * domain objects it creates. However, future iterations of this component should allow it to be more generic.
```
Comment (tresat) [inline subprojects/plugins/src/main/java/org/gradle/jvm/component/internal/DefaultJvmSoftwareComponent.java:78]: 👍

File: subprojects/plugins/src/main/java/org/gradle/api/plugins/JavaPlugin.java [hunk 3]
```diff
@@ -244,17 +244,16 @@ public void apply(final Project project) {
         project.getPluginManager().apply(JavaBasePlugin.class);
         project.getPluginManager().apply("org.gradle.jvm-test-suite");
 
-        JavaPluginExtension javaExtension = project.getExtensions().getByType(JavaPluginExtension.class);
-
         // Create the 'java' component.
-        JvmSoftwareComponentInternal component = objectFactory.newInstance(DefaultJvmSoftwareComponent.class, "java", javaExtension);
+        JvmSoftwareComponentInternal component = objectFactory.newInstance(DefaultJvmSoftwareComponent.class, "java", "main");
```
Comment (tresat) [inline subprojects/plugins/src/main/java/org/gradle/api/plugins/JavaPlugin.java:248]: 🧐 We should make static constants for these strings while reworking these classes.

**Review by tresat — APPROVED** [review]

File: subprojects/core/src/main/java/org/gradle/internal/service/scopes/ProjectScopeServices.java [hunk 4]
```diff
@@ -207,9 +207,9 @@ protected TaskContainerInternal createTaskContainerInternal(TaskStatistics taskS
         ).create();
     }
 
-    protected SoftwareComponentContainer createSoftwareComponentContainer(CollectionCallbackActionDecorator decorator, ObjectFactory objectFactory) {
-        Instantiator instantiator = get(Instantiator.class);
-        return instantiator.newInstance(DefaultSoftwareComponentContainer.class, instantiator, decorator, objectFactory);
+    protected SoftwareComponentContainer createSoftwareComponentContainer(Instantiator instantiator, InstantiatorFactory instantiatorFactory, ServiceRegistry services, CollectionCallbackActionDecorator decorator) {
+        Instantiator elementInstantiator = instantiatorFactory.decorate(services);
+        return elementInstantiator.newInstance(DefaultSoftwareComponentContainer.class, instantiator, elementInstantiator, decorator);
```
[the remark this answers is not in the document]
Comment (jvandort) [reply subprojects/core/src/main/java/org/gradle/internal/service/scopes/ProjectScopeServices.java:212]: The injected instantiator instance is very basic. It only contains the default services. The elements need a more advanced instantiator with access to more services, which is why we construct a new instantiator and register the injected services into it.

File: subprojects/plugins/src/main/java/org/gradle/jvm/component/internal/DefaultJvmSoftwareComponent.java (hunk 1 above)
Comment (jvandort) [reply subprojects/plugins/src/main/java/org/gradle/jvm/component/internal/DefaultJvmSoftwareComponent.java:227]: Leaving this TODO in for now as it does not need to be solved in this PR

**Comment by jvandort** [comment]
@bot-gradle test and merge

**Pull request merged**